docs(release-notes): write v3.1.0 and v3.2.0 release notes (ETL-1159) (#765)
glassflow/clickhouse-etl: docs/app/release-notes/v3.1.0/page.mdx

---
title: Release Notes v3.1.0
sidebarTitle: v3.1.0
icon: FileIcon
description: Release notes for the GlassFlow ClickHouse ETL v3.1.0. Backpressure across the data plane, configurable NATS stream caps, chunked NATS publish for bounded memory, and a one-shot data migration container.
---

# Release Notes v3.1.0

### Released: May 7, 2026

Version 3.1.0 introduces **first-class backpressure handling** across every stage of the data plane (ingestor, dedup, join, sink, and the OTLP receiver), surfaces a configurable **`MaxMsgs` cap on inter-stage NATS streams**, chunks the **NATS async publisher** to keep per-request memory bounded, and ships a new **one-shot data migration container** that turns v2-format pipeline configs into v3 on chart upgrade. This product release is accompanied by [**Helm chart v0.5.17**](https://github.com/glassflow/charts/releases/tag/glassflow-etl-0.5.17); use chart `v0.5.17` when installing or upgrading to v3.1.0.

## What's New

> ### 🔁 Backpressure across the data plane

Reviewer: ```suggestion
### Backpressure across the data plane
```

> Every stage that publishes to a JetStream stream now detects backpressure (the downstream stream hitting its `MaxMsgs` cap), retries with bounded backoff, and emits observability data so operators can see episodes as they happen.

Reviewer: ```suggestion
Every stage that publishes to a NATS JetStream stream now detects backpressure (the downstream stream hitting its `MaxMsgs` cap), retries with bounded backoff, and emits observability data so operators can see episodes as they happen.
```

- **Kafka ingestor** processes batches with backpressure awareness and re-queues failing records into a carry slice for the next batch attempt ([#708](https://github.com/glassflow/clickhouse-etl/pull/708)).
- **JetStream publisher** is now backpressure-aware on both the sync and async paths.
- **Dedup streaming component** retries with backoff when its output stream applies backpressure ([#715](https://github.com/glassflow/clickhouse-etl/pull/715)).
- **Join component** propagates backpressure on its output stream rather than silently looping ([#710](https://github.com/glassflow/clickhouse-etl/pull/710)).
- **OTLP receiver** applies backpressure to incoming requests when the downstream stream cannot accept publishes ([#711](https://github.com/glassflow/clickhouse-etl/pull/711)).
- **Observability** emits a `gfm_ingestor_backpressure_*` metric family covering active state, episode count, and per-episode duration.

> ### 🎚️ Configurable NATS stream cap (`MaxMsgs`)

Reviewer: ```suggestion
### Configurable NATS stream cap (`MaxMsgs`)
```

The inter-stage NATS streams now expose an explicit `MaxMsgs` cap on the pipeline definition, so operators can size the buffer to their throughput and latency requirements ([#701](https://github.com/glassflow/clickhouse-etl/pull/701)).

> - **Pipeline config.** `resources.nats.stream.maxMsgs` is the new knob. Defaults are computed automatically from the sink batch size, so most pipelines need no manual tuning.

Reviewer: ```suggestion
- **Pipeline config.** `resources.nats.stream.maxMsgs` is the new knob. Defaults are automatically computed from the sink batch size, so no manual tuning is necessary across most pipelines.
```

- **Backpressure trigger.** When a stream reaches `MaxMsgs`, the producing component enters a backpressure episode (see above) rather than dropping events.

> ### 🧱 NATS publisher chunking

Reviewer: ```suggestion
### NATS publisher chunking
```

The NATS async batch writer now chunks payloads so per-request memory is bounded regardless of upstream batch size ([#712](https://github.com/glassflow/clickhouse-etl/pull/712)).

- Each publish chunk is sized to a hard cap on memory rather than to the upstream batch.
- Prevents OOM scenarios when a single Kafka or OTLP request would otherwise produce an outsized NATS write.

> ### 🔄 Data migration container

Reviewer: ```suggestion
### Data migration container
```

A one-shot Kubernetes Job container migrates any persisted v2-format pipeline configs in your Postgres database to the v3 format.

- **What it does.** Runs once on chart install or upgrade, picks up v2 configs, applies the same transformation as the `migrate-preview` API, and writes the v3 result back.
- **What it does not do.** It does not invert `filter` expression semantics (that remains a human-judgment migration; see the [v3.0.0 release notes](/release-notes/v3.0.0#2-filter-expression-semantics-are-inverted)).
- **Image.** `glassflow/glassflow-etl-data-migration`, bundled in chart `v0.5.17`.

## Improvements

> ### 🔐 Security

Reviewer: ```suggestion
### Security
```

- Patched **CVE-2026-33815** in a transitive dependency ([#703](https://github.com/glassflow/clickhouse-etl/pull/703)).

> ## 🐛 Bug Fixes

Reviewer: ```suggestion
## Bug Fixes
```

- **v3 wizard alignment.** Pipeline-config generation in the UI wizard now aligns consistently with the v3 schema.

## Migration Notes

There are **no breaking changes** in v3.1.0 if you are already on v3.0.0. The data migration container handles any v2 pipeline configs still in Postgres on upgrade.

- **Helm chart.** Use chart `v0.5.17` for product `v3.1.0`. Charts `v0.5.18` through `v0.5.20` are chart-only patches that continue to bundle the `v3.1.0` binary (operational fixes, no product changes).
- **`MaxMsgs` default.** The newly computed default cap on inter-stage streams is sized from the sink's `max_batch_size`. If you previously relied on unbounded streams, set `resources.nats.stream.maxMsgs` explicitly on your pipelines.
- **Backpressure visibility.** If your dashboards alert on "consumer lag growing while no errors are emitted", you can now alert directly on `gfm_ingestor_backpressure_active == 1` for `5m` instead of inferring from rate-based queries.

## Try It Out

1. **Upgrade via Helm.** Deploy v3.1.0 using the [Kubernetes Helm charts](/installation/kubernetes) at chart version `v0.5.17`. Ensure your existing cluster is on v3.0.0 first.

> 2. **Verify the data migration.** Watch for the one-shot `glassflow-etl-data-migration` Job to complete; check Postgres for any pipeline rows whose persisted config is still in the v2 shape.

Reviewer: ```suggestion
2. **Verify the data migration.** Watch for the one-shot `glassflow-etl-data-migration` Job to complete; check Postgres for any pipeline rows where persisted config is still in the v2 shape.
```
